### ISSUE TITLE
False Positive: Type mismatch between list[T | U] and list[T | U] | list[T]

### ISSUE THREAD

**Issue by sahilmtayade** [astral-sh/ty#3051]
### Summary

**Description:**
The type checker incorrectly flags a type mismatch when a union contains the expected list type.

**Reproduction:**
[Playground Link](https://play.ty.dev/cdfdc51c-9f0f-499b-9122-9c5d2d42c0cf)

```python
a: list[list[float]]  = []

b = [[3,4,5], [3.3,2,3]]
c = [[1.2,3.3,4.4], [1,2,3]]
for p1 in b:
    for p2 in c:
        a.append(p1+p2)
print(a)
```

**Observed Error:**
`Argument to bound method `append` is incorrect: Expected `list[int | float]`, found `list[int | float] | list[int]` (invalid-argument-type) [Ln 9, Col 18]`

**Environment:**
*   **Extension:** `ty` (VS Code)
*   **Version:** 0.0.23

### Version

0.0.23

**Comment by sharkdp** [comment]
Thank you for reporting this.

We recently [changed type inference behavior](https://github.com/astral-sh/ruff/pull/23718) for list literals. The literal `[[3,4,5], [3.3,2,3]]` in your example contains one list of type `int` and one list of type `float` ([which we display as `int | float`](https://docs.astral.sh/ty/reference/typing-faq/#why-does-ty-show-int-float-when-i-annotate-something-as-float)). This is how you end up with the (somewhat unfortunate) inferred type of `list[list[int] | list[int | float]]` for `b` (and similar for `c`). That type later causes problems due to invariance of `list`.

I think this is an example where even https://github.com/astral-sh/ty/issues/1473 won't help, unless the `a.append(p1+p2)` call would also imply type context hints for `p1+p2` (and in effect for `p1` and `p2`, and therefore for `b` and `c`).

Other type checkers (mypy, pyrefly) will also emit an error for this example, or infer much less precise types like `list[Unknown]` for `b` and `c` (pyright). With no clear path on how we could improve ty's behavior here (short of reverting #23718), our current suggestion would be to [explicitly annotate `b` and `c`](https://play.ty.dev/96bd91bd-f84b-4c3d-a86c-c5363a45f4f5).

**Comment by ibraheemdev** [comment]
I think it would be reasonable for us to attempt to merge the element type of nested collections, and infer e.g., `[[1], ["2"]]` as `list[list[int | str]]` instead of `list[list[int] | list[str]]`. That does seem like a generally more useful type, and likely what was intended in most cases, though it's not clear whether this is true of all invariant generic instances, or just collection literals.

I do agree that tracking the uses of `b` and `c` through the iteration (as part of https://github.com/astral-sh/ty/issues/1473) seems very difficult.

**Comment by vitor-amartins** [comment]
I ran into what appears to be a very similar issue with the `tags` parameter of FastAPI's `APIRouter`, which is typed as `list[str | Enum] | None`.

Up to `ty` **0.0.20**, passing a `list[str]` worked without any diagnostics. Starting with **0.0.21**, I get the following error:

```text
error[invalid-argument-type]: Argument to bound method `post` is incorrect
   --> routes/well_routes.py:101:18
    |
100 |     @router.post('/{well_id}/archive', status_code=status.HTTP_200_OK, response_model=WellResponse,
101 |                  tags=tags, responses=get_responses_from_exceptions(validation_exceptions))
    |                  ^^^^^^^^^ Expected `list[str | Enum] | None`, found `list[str]`
```

This still reproduces with the latest `ty` version.

**Comment by carljm** [comment]
@vitor-amartins List is invariant, so it is correct to error when passing a `list[str]` to a `list[str | Enum]`, and ty has always errored on that (so do other type checkers). Probably in much older ty versions we were inferring your list as `list[Unknown]`, so this passed. Not sure exactly without seeing a full code example, but most likely wherever you are creating your `tags` list you should annotate it explicitly as `list[str | Enum]`.
